From 0462ff5564b713f246548d5dbbd97d1b4c95b5ca Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Tue, 31 Aug 2021 21:14:10 +0300
Subject: [PATCH] Merge pull request #3485 in SW_UGW/linux from
 feature/PONRTSYS-8636-move-functionality-of-pon_mcc_drv-to-standard-kernel-interfaces-part-1
 to master

* commit '296fc4eb29165fa206ff8db0c1eec64e876ebe6c':
  PONRTSYS-8636: Add dpm swdev_bridge_mcast_flood()
  PONRTSYS-8636: Add bridge IFLA_BR_MCAST_FLOOD
  PONRTSYS-8636: Add option to configure the SRC and DEST MCAST forwarding
  net: bridge: Add support for offloading port attributes
  net: switchdev: Add support for querying supported bridge flags by hardware
---
 include/linux/if_bridge.h    |   2 +
 include/net/datapath_inst.h  |   1 +
 include/net/switchdev.h      |   4 ++
 include/uapi/linux/if_link.h |  16 +++++
 net/bridge/br_netlink.c      | 139 +++++++++++++++++++++++++++++++++++++------
 net/bridge/br_private.h      |  11 ++++
 net/bridge/br_switchdev.c    |  44 ++++++++++++++
 net/bridge/br_sysfs_if.c     |   4 ++
 8 files changed, 202 insertions(+), 19 deletions(-)

diff --git a/include/linux/if_bridge.h b/include/linux/if_bridge.h
index e9732e1ebccc..e90d5384c092 100644
--- a/include/linux/if_bridge.h
+++ b/include/linux/if_bridge.h
@@ -47,6 +47,8 @@ struct br_ip_list {
 #define BR_PROXYARP_WIFI	BIT(10)
 #define BR_MCAST_FLOOD		BIT(11)
 #define BR_ISOLATED		BIT(13)
+#define BR_MCAST_DEST_LOOKUP		BIT(14)
+#define BR_MCAST_SRC_LOOKUP		BIT(15)
 
 #define BR_DEFAULT_AGEING_TIME	(300 * HZ)
 
diff --git a/include/net/datapath_inst.h b/include/net/datapath_inst.h
index 942d59f6e181..3bfbcc5b0a21 100644
--- a/include/net/datapath_inst.h
+++ b/include/net/datapath_inst.h
@@ -172,6 +172,7 @@ struct inst_info {
 	int (*dp_mac_set)(int bport, int fid, int inst, u8 *addr);
 	int (*dp_mac_reset)(int bport, int fid, int inst, u8 *addr);
 	int (*dp_cfg_vlan)(int inst, int vap, int ep);
+	int (*swdev_bridge_mcast_flood)(int inst, int br_id, bool activate);
 #endif
 	int (*dp_tc_vlan_set)(struct core_ops *ops, struct dp_tc_vlan *vlan,
 			      struct dp_tc_vlan_info *info,
diff --git a/include/net/switchdev.h b/include/net/switchdev.h
index 2fa2cbf7c10e..bff2176cacdd 100644
--- a/include/net/switchdev.h
+++ b/include/net/switchdev.h
@@ -47,8 +47,10 @@ enum switchdev_attr_id {
 	SWITCHDEV_ATTR_ID_PORT_STP_STATE,
 	SWITCHDEV_ATTR_ID_PORT_BRIDGE_FLAGS,
 	SWITCHDEV_ATTR_ID_PORT_LEARNING_LIMIT,
+	SWITCHDEV_ATTR_ID_PORT_BRIDGE_FLAGS_SUPPORT,
 	SWITCHDEV_ATTR_ID_BRIDGE_AGEING_TIME,
 	SWITCHDEV_ATTR_ID_BRIDGE_VLAN_FILTERING,
+	SWITCHDEV_ATTR_ID_BRIDGE_MCAST_FLOOD,
 };
 
 struct switchdev_attr {
@@ -62,8 +64,10 @@ struct switchdev_attr {
 		u8 stp_state;				/* PORT_STP_STATE */
 		unsigned long brport_flags;		/* PORT_BRIDGE_FLAGS */
 		int learning_limit;			/* PORT_LEARNING_LIMIT */
+		unsigned long brport_flags_support;	/* PORT_BRIDGE_FLAGS_SUPPORT */
 		clock_t ageing_time;			/* BRIDGE_AGEING_TIME */
 		bool vlan_filtering;			/* BRIDGE_VLAN_FILTERING */
+		enum br_mcast_flood_mode mcast_flood;		/* BRIDGE_MCAST_FLOOD */
 	} u;
 };
 
diff --git a/include/uapi/linux/if_link.h b/include/uapi/linux/if_link.h
index 4d696b6b3815..dfcf290aa76a 100644
--- a/include/uapi/linux/if_link.h
+++ b/include/uapi/linux/if_link.h
@@ -275,11 +275,25 @@ enum {
 	IFLA_BR_PAD,
 	IFLA_BR_VLAN_STATS_ENABLED,
 	IFLA_BR_MCAST_STATS_ENABLED,
+	IFLA_BR_MCAST_IGMP_VERSION,
+	IFLA_BR_MCAST_MLD_VERSION,
+	IFLA_BR_VLAN_STATS_PER_PORT,
+	IFLA_BR_MULTI_BOOLOPT,
+	IFLA_BR_MCAST_FLOOD,
 	__IFLA_BR_MAX,
 };
 
 #define IFLA_BR_MAX	(__IFLA_BR_MAX - 1)
 
+enum br_mcast_flood_mode {
+	/** Packet is flooded to port members of ingress bridge port */
+	BR_MCAST_FLOOD_ALL,
+	/** Packet is dscarded */
+	BR_MCAST_FLOOD_DISCARD,
+	/** Packet is forwarded to logical port 0 CTP port 0 bridge port 0 */
+	BR_MCAST_FLOOD_CPU,
+};
+
 struct ifla_bridge_id {
 	__u8	prio[2];
 	__u8	addr[6]; /* ETH_ALEN */
@@ -329,6 +343,8 @@ enum {
 	IFLA_BRPORT_ISOLATED,
 	IFLA_BRPORT_BACKUP_PORT,
 	IFLA_BRPORT_LEARNING_LIMIT,
+	IFLA_BRPORT_DEST_LOOKUP,
+	IFLA_BRPORT_SRC_LOOKUP,
 	__IFLA_BRPORT_MAX
 };
 #define IFLA_BRPORT_MAX (__IFLA_BRPORT_MAX - 1)
diff --git a/net/bridge/br_netlink.c b/net/bridge/br_netlink.c
index 1592be9c419e..d466a67831d4 100644
--- a/net/bridge/br_netlink.c
+++ b/net/bridge/br_netlink.c
@@ -144,6 +144,8 @@ static inline size_t br_port_info_size(void)
 		+ nla_total_size(sizeof(u8))	/* IFLA_BRPORT_MULTICAST_ROUTER */
 #endif
 		+ nla_total_size(4)	/* IFLA_BRPORT_LEARNING_LIMIT */
+		+ nla_total_size(sizeof(u8))	/* IFLA_BRPORT_DEST_LOOKUP */
+		+ nla_total_size(sizeof(u8))	/* IFLA_BRPORT_SRC_LOOKUP */
 		+ 0;
 }
 
@@ -196,6 +198,8 @@ static int br_port_fill_attrs(struct sk_buff *skb,
 		       p->topology_change_ack) ||
 	    nla_put_u8(skb, IFLA_BRPORT_CONFIG_PENDING, p->config_pending) ||
 	    nla_put_u32(skb, IFLA_BRPORT_LEARNING_LIMIT, p->learning_limit) ||
+	    nla_put_u8(skb, IFLA_BRPORT_DEST_LOOKUP, !!(p->flags & BR_MCAST_DEST_LOOKUP)) ||
+	    nla_put_u8(skb, IFLA_BRPORT_SRC_LOOKUP, !!(p->flags & BR_MCAST_SRC_LOOKUP)) ||
 	    nla_put_u8(skb, IFLA_BRPORT_ISOLATED, !!(p->flags & BR_ISOLATED)))
 		return -EMSGSIZE;
 
@@ -593,6 +597,8 @@ static const struct nla_policy br_port_policy[IFLA_BRPORT_MAX + 1] = {
 	[IFLA_BRPORT_MULTICAST_ROUTER] = { .type = NLA_U8 },
 	[IFLA_BRPORT_LEARNING_LIMIT] = { .type = NLA_S32 },
 	[IFLA_BRPORT_ISOLATED]	= { .type = NLA_U8 },
+	[IFLA_BRPORT_DEST_LOOKUP]	= { .type = NLA_U8 },
+	[IFLA_BRPORT_SRC_LOOKUP]	= { .type = NLA_U8 },
 };
 
 /* Change the state of the port and notify spanning tree */
@@ -649,17 +655,51 @@ static int br_set_learning_limit(struct net_bridge_port *p, int learning_limit)
 	return 0;
 }
 
+/* Change the state of the port and notify spanning tree */
+static int br_set_mcast_flood(struct net_bridge *br, enum br_mcast_flood_mode mcast_flood)
+{
+	struct switchdev_attr attr = {
+		.orig_dev = br->dev,
+		.id = SWITCHDEV_ATTR_ID_BRIDGE_MCAST_FLOOD,
+		.flags = SWITCHDEV_F_SKIP_EOPNOTSUPP,
+		.u.mcast_flood = mcast_flood,
+	};
+	int err;
+
+	br->mcast_flood = mcast_flood;
+
+	err = switchdev_port_attr_set(br->dev, &attr);
+	if (err && err != -EOPNOTSUPP)
+		br_warn(br, "error setting multicast flooding on bridge\n");
+	else
+		br_debug(br, "multicast flooding set to %i bridge\n",
+			 br->mcast_flood);
+
+	br_configuration_update(br);
+	return 0;
+}
+
 /* Set/clear or port flags based on attribute */
-static void br_set_port_flag(struct net_bridge_port *p, struct nlattr *tb[],
-			   int attrtype, unsigned long mask)
+static int br_set_port_flag(struct net_bridge_port *p, struct nlattr *tb[],
+			    int attrtype, unsigned long mask)
 {
-	if (tb[attrtype]) {
-		u8 flag = nla_get_u8(tb[attrtype]);
-		if (flag)
-			p->flags |= mask;
-		else
-			p->flags &= ~mask;
-	}
+	unsigned long flags;
+	int err;
+
+	if (!tb[attrtype])
+		return 0;
+
+	if (nla_get_u8(tb[attrtype]))
+		flags = p->flags | mask;
+	else
+		flags = p->flags & ~mask;
+
+	err = br_switchdev_set_port_flag(p, flags, mask);
+	if (err)
+		return err;
+
+	p->flags = flags;
+	return 0;
 }
 
 /* Process bridge protocol info on port */
@@ -668,16 +708,53 @@ static int br_setport(struct net_bridge_port *p, struct nlattr *tb[])
 	int err;
 	unsigned long old_flags = p->flags;
 
-	br_set_port_flag(p, tb, IFLA_BRPORT_MODE, BR_HAIRPIN_MODE);
-	br_set_port_flag(p, tb, IFLA_BRPORT_GUARD, BR_BPDU_GUARD);
-	br_set_port_flag(p, tb, IFLA_BRPORT_FAST_LEAVE, BR_MULTICAST_FAST_LEAVE);
-	br_set_port_flag(p, tb, IFLA_BRPORT_PROTECT, BR_ROOT_BLOCK);
-	br_set_port_flag(p, tb, IFLA_BRPORT_LEARNING, BR_LEARNING);
-	br_set_port_flag(p, tb, IFLA_BRPORT_UNICAST_FLOOD, BR_FLOOD);
-	br_set_port_flag(p, tb, IFLA_BRPORT_MCAST_FLOOD, BR_MCAST_FLOOD);
-	br_set_port_flag(p, tb, IFLA_BRPORT_PROXYARP, BR_PROXYARP);
-	br_set_port_flag(p, tb, IFLA_BRPORT_PROXYARP_WIFI, BR_PROXYARP_WIFI);
-	br_set_port_flag(p, tb, IFLA_BRPORT_ISOLATED, BR_ISOLATED);
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_MODE, BR_HAIRPIN_MODE);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_GUARD, BR_BPDU_GUARD);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_FAST_LEAVE, BR_MULTICAST_FAST_LEAVE);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_PROTECT, BR_ROOT_BLOCK);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_LEARNING, BR_LEARNING);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_UNICAST_FLOOD, BR_FLOOD);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_MCAST_FLOOD, BR_MCAST_FLOOD);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_PROXYARP, BR_PROXYARP);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_PROXYARP_WIFI, BR_PROXYARP_WIFI);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_ISOLATED, BR_ISOLATED);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_DEST_LOOKUP, BR_MCAST_DEST_LOOKUP);
+	if (err)
+		return err;
+
+	err = br_set_port_flag(p, tb, IFLA_BRPORT_SRC_LOOKUP, BR_MCAST_SRC_LOOKUP);
+	if (err)
+		return err;
 
 	if (tb[IFLA_BRPORT_COST]) {
 		err = br_stp_set_path_cost(p, nla_get_u32(tb[IFLA_BRPORT_COST]));
@@ -831,6 +908,17 @@ static int br_validate(struct nlattr *tb[], struct nlattr *data[])
 	}
 #endif
 
+	if (data[IFLA_BR_MCAST_FLOOD]) {
+		switch (nla_get_u8(data[IFLA_BR_MCAST_FLOOD])) {
+		case BR_MCAST_FLOOD_ALL:
+		case BR_MCAST_FLOOD_DISCARD:
+		case BR_MCAST_FLOOD_CPU:
+			break;
+		default:
+			return -EPROTONOSUPPORT;
+		}
+	}
+
 	return 0;
 }
 
@@ -897,6 +985,7 @@ static const struct nla_policy br_policy[IFLA_BR_MAX + 1] = {
 	[IFLA_BR_VLAN_DEFAULT_PVID] = { .type = NLA_U16 },
 	[IFLA_BR_VLAN_STATS_ENABLED] = { .type = NLA_U8 },
 	[IFLA_BR_MCAST_STATS_ENABLED] = { .type = NLA_U8 },
+	[IFLA_BR_MCAST_FLOOD] = { .type = NLA_U8 },
 };
 
 static int br_changelink(struct net_device *brdev, struct nlattr *tb[],
@@ -1109,6 +1198,15 @@ static int br_changelink(struct net_device *brdev, struct nlattr *tb[],
 		br->multicast_stats_enabled = !!mcast_stats;
 	}
 #endif
+	if (data[IFLA_BR_MCAST_FLOOD]) {
+		enum br_mcast_flood_mode mcast_flood;
+		
+		mcast_flood = nla_get_u8(data[IFLA_BR_MCAST_FLOOD]);
+
+		err = br_set_mcast_flood(br, mcast_flood);
+		if (err)
+			return err;
+	}
 #if IS_ENABLED(CONFIG_BRIDGE_NETFILTER)
 	if (data[IFLA_BR_NF_CALL_IPTABLES]) {
 		u8 val = nla_get_u8(data[IFLA_BR_NF_CALL_IPTABLES]);
@@ -1187,6 +1285,7 @@ static size_t br_get_size(const struct net_device *brdev)
 	       nla_total_size(sizeof(u8)) +     /* IFLA_BR_MCAST_QUERY_USE_IFADDR */
 	       nla_total_size(sizeof(u8)) +     /* IFLA_BR_MCAST_QUERIER */
 	       nla_total_size(sizeof(u8)) +     /* IFLA_BR_MCAST_STATS_ENABLED */
+	       nla_total_size(sizeof(u8)) +     /* IFLA_BR_MCAST_FLOOD */
 	       nla_total_size(sizeof(u32)) +    /* IFLA_BR_MCAST_HASH_ELASTICITY */
 	       nla_total_size(sizeof(u32)) +    /* IFLA_BR_MCAST_HASH_MAX */
 	       nla_total_size(sizeof(u32)) +    /* IFLA_BR_MCAST_LAST_MEMBER_CNT */
@@ -1266,6 +1365,8 @@ static int br_fill_info(struct sk_buff *skb, const struct net_device *brdev)
 	    nla_put_u8(skb, IFLA_BR_MCAST_QUERIER, br->multicast_querier) ||
 	    nla_put_u8(skb, IFLA_BR_MCAST_STATS_ENABLED,
 		       br->multicast_stats_enabled) ||
+	    nla_put_u8(skb, IFLA_BR_MCAST_FLOOD,
+		       br->mcast_flood) ||
 	    nla_put_u32(skb, IFLA_BR_MCAST_HASH_ELASTICITY,
 			br->hash_elasticity) ||
 	    nla_put_u32(skb, IFLA_BR_MCAST_HASH_MAX, br->hash_max) ||
diff --git a/net/bridge/br_private.h b/net/bridge/br_private.h
index 2d6792c83502..cd35fd06fff8 100644
--- a/net/bridge/br_private.h
+++ b/net/bridge/br_private.h
@@ -383,6 +383,7 @@ struct net_bridge {
 	unsigned char			topology_change;
 	unsigned char			topology_change_detected;
 
+	enum br_mcast_flood_mode	mcast_flood;
 #ifdef CONFIG_BRIDGE_IGMP_SNOOPING
 	unsigned char			multicast_router;
 
@@ -1146,6 +1147,9 @@ void nbp_switchdev_frame_mark(const struct net_bridge_port *p,
 			      struct sk_buff *skb);
 bool nbp_switchdev_allowed_egress(const struct net_bridge_port *p,
 				  const struct sk_buff *skb);
+int br_switchdev_set_port_flag(struct net_bridge_port *p,
+			       unsigned long flags,
+			       unsigned long mask);
 #else
 static inline int nbp_switchdev_mark_set(struct net_bridge_port *p)
 {
@@ -1162,6 +1166,13 @@ static inline bool nbp_switchdev_allowed_egress(const struct net_bridge_port *p,
 {
 	return true;
 }
+
+static inline int br_switchdev_set_port_flag(struct net_bridge_port *p,
+					     unsigned long flags,
+					     unsigned long mask)
+{
+	return 0;
+}
 #endif /* CONFIG_NET_SWITCHDEV */
 
 #endif
diff --git a/net/bridge/br_switchdev.c b/net/bridge/br_switchdev.c
index f4097b900de1..f8cc7121f40f 100644
--- a/net/bridge/br_switchdev.c
+++ b/net/bridge/br_switchdev.c
@@ -55,3 +55,47 @@ bool nbp_switchdev_allowed_egress(const struct net_bridge_port *p,
 	return !skb->offload_fwd_mark ||
 	       BR_INPUT_SKB_CB(skb)->offload_fwd_mark != p->offload_fwd_mark;
 }
+
+/* Flags that can be offloaded to hardware */
+#define BR_PORT_FLAGS_HW_OFFLOAD (BR_LEARNING | BR_FLOOD | \
+				  BR_MCAST_FLOOD | BR_MCAST_DEST_LOOKUP | \
+				  BR_MCAST_SRC_LOOKUP)
+
+int br_switchdev_set_port_flag(struct net_bridge_port *p,
+			       unsigned long flags,
+			       unsigned long mask)
+{
+	struct switchdev_attr attr = {
+		.orig_dev = p->dev,
+		.id = SWITCHDEV_ATTR_ID_PORT_BRIDGE_FLAGS_SUPPORT,
+	};
+	int err;
+
+	if (mask & ~BR_PORT_FLAGS_HW_OFFLOAD)
+		return 0;
+
+	err = switchdev_port_attr_get(p->dev, &attr);
+	if (err == -EOPNOTSUPP)
+		return 0;
+	if (err)
+		return err;
+
+	/* Check if specific bridge flag attribute offload is supported */
+	if (!(attr.u.brport_flags_support & mask)) {
+		br_warn(p->br, "bridge flag offload is not supported %u(%s)\n",
+			(unsigned int)p->port_no, p->dev->name);
+		return -EOPNOTSUPP;
+	}
+
+	attr.id = SWITCHDEV_ATTR_ID_PORT_BRIDGE_FLAGS;
+	attr.flags = SWITCHDEV_F_DEFER;
+	attr.u.brport_flags = flags;
+	err = switchdev_port_attr_set(p->dev, &attr);
+	if (err) {
+		br_warn(p->br, "error setting offload flag on port %u(%s)\n",
+			(unsigned int)p->port_no, p->dev->name);
+		return err;
+	}
+
+	return 0;
+}
diff --git a/net/bridge/br_sysfs_if.c b/net/bridge/br_sysfs_if.c
index 9cf2667bb9af..906056013cb4 100644
--- a/net/bridge/br_sysfs_if.c
+++ b/net/bridge/br_sysfs_if.c
@@ -180,6 +180,8 @@ BRPORT_ATTR_FLAG(proxyarp, BR_PROXYARP);
 BRPORT_ATTR_FLAG(proxyarp_wifi, BR_PROXYARP_WIFI);
 BRPORT_ATTR_FLAG(multicast_flood, BR_MCAST_FLOOD);
 BRPORT_ATTR_FLAG(isolated, BR_ISOLATED);
+BRPORT_ATTR_FLAG(mcast_dest_lookup, BR_MCAST_DEST_LOOKUP);
+BRPORT_ATTR_FLAG(mcast_src_lookup, BR_MCAST_SRC_LOOKUP);
 
 #ifdef CONFIG_BRIDGE_IGMP_SNOOPING
 static ssize_t show_multicast_router(struct net_bridge_port *p, char *buf)
@@ -227,6 +229,8 @@ static const struct brport_attribute *brport_attrs[] = {
 	&brport_attr_proxyarp_wifi,
 	&brport_attr_multicast_flood,
 	&brport_attr_isolated,
+	&brport_attr_mcast_dest_lookup,
+	&brport_attr_mcast_src_lookup,
 	NULL
 };
 
